File: trading_py/scripts/finnhub_fetch.py
#!/usr/bin/env python3
"""Fetch historical data from Finnhub and save CSV.
Usage: FINNHUB_API_KEY=... python finnhub_fetch.py --symbol AAPL --resolution D --outfile out/AAPL.csv
"""
import argparse
import os
from pathlib import Path
from trading_py.connectors.finnhub_connector import FinnhubClient

def main():
    p = argparse.ArgumentParser()
    p.add_argument('--symbol', required=True)
    p.add_argument('--resolution', default='D')
    p.add_argument('--from_ts', type=int, default=None)
    p.add_argument('--to_ts', type=int, default=None)
    p.add_argument('--outfile', default=None)
    args = p.parse_args()

    client = FinnhubClient()
    df = client.get_candles(args.symbol, resolution=args.resolution, _from=args.from_ts, to=args.to_ts)
    out = args.outfile or f"trading-py/output/{args.symbol.replace('/','_')}_finnhub.csv"
    Path(os.path.dirname(out) or '.').mkdir(parents=True, exist_ok=True)
    df.to_csv(out)
    print(f"Saved {len(df)} rows to {out}")

if __name__ == '__main__':
    main()

File: trading-stack/finnhub_fetch.js
#!/usr/bin/env node
// usage: node trading-stack/finnhub_fetch.js --symbol=AAPL --apikey=YOUR_KEY --resolution=D --days=365 --out=out.csv
const axios = require('axios');
const fs = require('fs');
const argv = Object.fromEntries(process.argv.slice(2).map(s=>s.replace(/^--/,'').split('=')));
const symbol = argv.symbol || 'AAPL';
const apiKey = argv.apikey || process.env.FINNHUB_API_KEY;
const resolution = argv.resolution || 'D';
const days = parseInt(argv.days||'365',10);
const out = argv.out || `trading-stack/${symbol}_finnhub.csv`;
if(!apiKey){ console.error('Missing api key (use --apikey or set FINNHUB_API_KEY)'); process.exit(1); }
(async()=>{
  try{
    const to = Math.floor(Date.now()/1000);
    const from = to - days*24*3600;
    const url = 'https://finnhub.io/api/v1/stock/candle';
    const r = await axios.get(url, { params:{ symbol, resolution, from, to, token: apiKey }, timeout:30000 });
    const j = r.data;
    if(j.s !== 'ok'){ console.error('Finnhub error', j); process.exit(2); }
    const rows = ['date,open,high,low,close,volume'];
    for(let i=0;i<j.t.length;i++){
      const dt = new Date(j.t[i]*1000).toISOString();
      rows.push([dt, j.o[i], j.h[i], j.l[i], j.c[i], j.v[i]].join(','));
    }
    fs.mkdirSync(require('path').dirname(out), { recursive:true });
    fs.writeFileSync(out, rows.join('\n'));
    console.log('Saved', out, 'rows=', j.t.length);
  }catch(e){
    console.error('Fetch failed', e.message || e);
    process.exit(3);
  }
})();

